R/board_gcs.R

```diff
@@ -0,0 +1,262 @@
+    check_pin_exists(board, name)
+    gcs_delete_dir(board, name)
+  }
+  invisible(board)
+}
+
+#' @export
+pin_versions.pins_board_gcs <- function(board, name, ...) {
+  check_pin_exists(board, name)
+  resp <- googleCloudStorageR::gcs_list_objects(
+    bucket = board$bucket,
+    prefix = paste0(board$prefix, name)
+  )
+  paths <- fs::path_split(unique(fs::path_dir(resp$name)))
+  version_from_path(map_chr(paths, ~ .x[[length(.x)]]))
+}
+
+#' @export
+pin_version_delete.pins_board_gcs <- function(board, name, version, ...) {
+  gcs_delete_dir(board, fs::path(name, version))
+}
+
+#' @export
+pin_meta.pins_board_gcs <- function(board, name, version = NULL, ...) {
+  withr::local_options(list(googleAuthR.verbose = 4))
+  check_pin_exists(board, name)
+  version <- check_pin_version(board, name, version)
+  metadata_blob <- fs::path(name, version, "data.txt")
+
+  if (!gcs_file_exists(board, metadata_blob)) {
+    abort_pin_version_missing(version)
+  }
+
+  path_version <- fs::path(board$cache, name, version)
+  fs::dir_create(path_version)
+  gcs_download(board, metadata_blob)
+  local_meta(
+    read_meta(fs::path(board$cache, name, version)),
+    name = name,
+    dir = path_version,
+    version = version
+  )
+}
+
+#' @export
+pin_fetch.pins_board_gcs <- function(board, name, version = NULL, ...) {
+  withr::local_options(list(googleAuthR.verbose = 4))
+  meta <- pin_meta(board, name, version = version)
+  cache_touch(board, meta)
+
+  for (file in meta$file) {
+    key <- fs::path(name, meta$local$version, file)
+    gcs_download(board, key)
+  }
+
+  meta
+}
+
+#' @export
+pin_store.pins_board_gcs <- function(board, name, paths, metadata,
+                                     versioned = NULL, x = NULL, ...) {
+  withr::local_options(list(googleAuthR.verbose = 4))
+  ellipsis::check_dots_used()
+  check_name(name)
+  version <- version_setup(board, name, version_name(metadata), versioned = versioned)
+  version_dir <- fs::path(name, version)
+  gcs_upload_yaml(
+    board,
+    fs::path(paste0(board$prefix, version_dir), "data.txt"),
+    metadata
+  )
+
+  for (path in paths) {
+    googleCloudStorageR::gcs_upload(
+      file = path,
+      bucket = board$bucket,
+      name = fs::path(paste0(board$prefix, version_dir), fs::path_file(path)),
+      ...
+    )
+  }
+
+  name
+}
+
+#' @rdname board_deparse
+#' @export
+board_deparse.pins_board_gcs <- function(board, ...) {
+  bucket <- check_board_deparse(board, "bucket")
+  expr(board_gcs(!!bucket, prefix = !!board$prefix))
+}
+
+#' @rdname required_pkgs.pins_board
+#' @export
+required_pkgs.pins_board_gcs <- function(x, ...) {
+  ellipsis::check_dots_empty()
+  "googleCloudStorageR"
+}
+
+# Helpers -----------------------------------------------------------------
+
+gcs_delete_dir <- function(board, dir) {
+  resp <- googleCloudStorageR::gcs_list_objects(
+    bucket = board$bucket,
+    prefix = paste0(board$prefix, dir, "/")
+  )
+
+  if (nrow(resp) == 0) {
```

Comment:
I think this is probably not needed unless when `nrow()` is a 0, there's no `name` column.

Reply:
Yep, that is the situation:

``` r
library(pins)
googleCloudStorageR::gcs_auth("~/google-pins.json")
b <- board_gcs("pins-testing")
resp <- googleCloudStorageR::gcs_list_objects(
  bucket = b$bucket,
  prefix = paste0(b$prefix, "cats-and-dogs", "/")
)
#> ℹ 2023-01-12 10:50:27 > No objects found
resp
#> data frame with 0 columns and 0 rows
```

<sup>Created on 2023-01-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>